### PR TITLE
core: implement a method to dump user ta's runtime status

### PR DESCRIPTION
**Pull request by weizhaojiang**

### CODE REVIEW

File: core/kernel/tee_ta_manager.c [hunk 1]
```diff
@@ -823,6 +842,188 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
...
+		stats->panicked = dump_ctx[i].panicked;
+		stats->sess_num = dump_ctx[i].sess_num;
+
+		/* find a session from dump context */
+		sess = NULL;
+		for (j = 0; j < dump_ctx[i].sess_num; j++) {
+			uint32_t sess_id = dump_ctx[i].sess_id[j];
+
+			sess = tee_ta_get_session(sess_id,
+						  true,
```
Comment (jenswi-linaro) [inline core/kernel/tee_ta_manager.c:942]: Please fold up these two lines.

File: core/kernel/tee_ta_manager.c [hunk 2]
```diff
@@ -823,6 +842,188 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
...
+						  open_sessions);
+			if (sess)
```
Comment (jenswi-linaro) [inline core/kernel/tee_ta_manager.c:944]: Better to test this together with `j < dump_ctx[i].sess_num` above in the condition part of the for loop.

**Review by jenswi-linaro — COMMENTED** [review]
With my comments addressed please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: lib/libutee/include/utee_types.h [hunk 3]
```diff
@@ -19,6 +19,16 @@ enum utee_entry_func {
 	UTEE_ENTRY_FUNC_OPEN_SESSION = 0,
 	UTEE_ENTRY_FUNC_CLOSE_SESSION,
 	UTEE_ENTRY_FUNC_INVOKE_COMMAND,
+	/*
+	 * UTEE_ENTRY_FUNC_DUMP_MEMSTATS
+	 * [out] value[0].a  Byte currently allocated
```
Comment (etienne-lms) [inline lib/libutee/include/utee_types.h:24]: ditto (my fault)
"Byte size currently ..."
or
"Bytes currently ..."

File: core/kernel/tee_ta_manager.c (hunk 2 above)
Comment (etienne-lms) [reply core/kernel/tee_ta_manager.c:944]: maybe like this:
```c
		/* Find a session from dump context */
		for (j = 0, sess = NULL; j < dump_ctx[i].sess_num && !sess;j++)
			sess = tee_ta_get_session(dump_ctx[i].sess_id[j], true,
						   open_sessions);
```

(by the way, nitpicking on consistency, prefer an uppercase leter for inline comment first word fisrt letter, here:
```patch
-		/* find a session from dump context */
+		/* Find a session from dump context */
```

File: core/kernel/tee_ta_manager.c [hunk 4]
```diff
@@ -823,6 +842,188 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
 	return res;
 }
 
+#if defined(CFG_TA_STATS)
+static TEE_Result dump_ta_memstats(struct tee_ta_session *s,
+				   uint32_t cancel_req_to,
```
Comment (etienne-lms) [inline core/kernel/tee_ta_manager.c:847]: Do we really need `cancel_req_to` argument here?

**Review by etienne-lms — COMMENTED** [review]
minor stuff

File: core/kernel/tee_ta_manager.c (hunk 1 above)
Comment (weizhaojiang) [reply core/kernel/tee_ta_manager.c:942]: done

File: core/kernel/tee_ta_manager.c (hunk 2 above)
Comment (weizhaojiang) [reply core/kernel/tee_ta_manager.c:944]: done

File: core/pta/stats.c [hunk 5]
```diff
@@ -21,6 +21,22 @@
 #define STATS_CMD_PAGER_STATS		0
 #define STATS_CMD_ALLOC_STATS		1
 #define STATS_CMD_MEMLEAK_STATS		2
+/*
+ * UTEE_ENTRY_FUNC_DUMP_MEMSTATS
+ * [out]    memref[0]        Array of context information of loaded TAs
+ *
+ * Each cell of the TA information information array contains:
+ * TEE_UUID    TA UUID
+ * uint32_t    Non zero is TA panicked, 0 otherwise
```
[the remark this answers is not in the document]
Comment (weizhaojiang) [reply core/pta/stats.c:30]: done

File: core/pta/stats.c [hunk 6]
```diff
@@ -21,6 +21,22 @@
...
+ * uint32_t    Number of sessions opened by the TA
+ * uint32_t    Byte currently allocated in TA heap
```
[the remark this answers is not in the document]
Comment (weizhaojiang) [reply core/pta/stats.c:32]: done

File: core/pta/stats.c [hunk 7]
```diff
@@ -137,6 +153,32 @@ static TEE_Result get_memleak_stats(uint32_t type,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result get_user_ta_stats(uint32_t type,
+				    TEE_Param p[TEE_NUM_PARAMS] __unused)
```
[the remark this answers is not in the document]
Comment (weizhaojiang) [reply core/pta/stats.c:157]: done

File: lib/libutee/arch/arm/user_ta_entry.c [hunk 8]
```diff
@@ -374,6 +374,37 @@ static TEE_Result entry_invoke_command(unsigned long session_id,
...
+
+	from_utee_params(params, &param_types, up);
+	ta_header_save_params(param_types, params);
+
+	if (TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+			    TEE_PARAM_TYPE_VALUE_OUTPUT,
+			    TEE_PARAM_TYPE_VALUE_OUTPUT,
+			    TEE_PARAM_TYPE_NONE) != param_types) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
```
[the remark this answers is not in the document]
Comment (weizhaojiang) [reply lib/libutee/arch/arm/user_ta_entry.c:393]: done

File: core/pta/stats.c [hunk 9]
```diff
@@ -137,6 +153,32 @@ static TEE_Result get_memleak_stats(uint32_t type,
...
+{
+	uint32_t res = TEE_SUCCESS;
+
+	/*
+	 * p[0].memref.buffer = output buffer to struct malloc_stats.
+	 */
```
[the remark this answers is not in the document]
Comment (weizhaojiang) [reply core/pta/stats.c:163]: done

File: lib/libutee/include/utee_types.h (hunk 3 above)
Comment (weizhaojiang) [reply lib/libutee/include/utee_types.h:24]: done

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

File: lib/libutee/include/utee_types.h (hunk 3 above)
Comment (etienne-lms) [inline lib/libutee/include/utee_types.h:24]: "Byte size currently ..."
or
"Bytes currently ..."

File: core/pta/stats.c [hunk 10]
```diff
@@ -137,6 +153,28 @@ static TEE_Result get_memleak_stats(uint32_t type,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result get_user_ta_stats(uint32_t type,
+				    TEE_Param p[TEE_NUM_PARAMS] __unused)
```
Comment (etienne-lms) [inline core/pta/stats.c:157]: s/__unused/__maybe_unused/

File: core/kernel/tee_ta_manager.c [hunk 11]
```diff
@@ -823,6 +842,185 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
...
+		       sizeof(dump_ctx[i].uuid));
+		stats->panicked = dump_ctx[i].panicked;
+		stats->sess_num = dump_ctx[i].sess_num;
+
+		/* Find a session from dump context */
+		sess = NULL;
+		for (j = 0; j < dump_ctx[i].sess_num; j++) {
+			uint32_t sess_id = dump_ctx[i].sess_id[j];
+
+			sess = tee_ta_get_session(sess_id, true,open_sessions);
```
[the remark this answers is not in the document]
Comment (weizhaojiang) [reply core/kernel/tee_ta_manager.c:940]: done

File: lib/libutee/include/utee_types.h (hunk 3 above)
Comment (weizhaojiang) [reply lib/libutee/include/utee_types.h:24]: done

File: mk/config.mk [hunk 12]
```diff
@@ -946,3 +946,10 @@ CFG_CORE_PREALLOC_EL0_TBLS ?= n
 ifeq (y-y,$(CFG_CORE_PREALLOC_EL0_TBLS)-$(CFG_WITH_PAGER))
 $(error "CFG_WITH_PAGER can't support CFG_CORE_PREALLOC_EL0_TBLS")
 endif
+
+# User TA runtime context dump.
+# When this option is enabled, OP-TEE provides a debug method for
+# developer to dump user TA's runtime context, including TA's heap stats.
+# Developer can open a stats PTA session and then invoke command
+# STATS_CMD_TA_STATS to get the context of loaded TAs.
```
Comment (jforissier) [inline mk/config.mk:954]: It would be nice to extend `xtest` for that: https://github.com/OP-TEE/optee_test/blob/master/host/xtest/stats.c.

Comment (weizhaojiang) [reply mk/config.mk:954]: Sure. I actually have the similar sample code in my local. Let me submit it in a new PR right after the current one get merged.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by weizhaojiang** [comment]
> `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

thanks for the review.

**Comment by jforissier** [comment]
Merged with small edits to the commit messages. Thanks!

**Pull request closed without merging**